File: src/app/groups/annealing/move.ts
import { Solution, Restriction } from "./types"; // assuming you've defined these types elsewhere
import { deepCopy, randomIntFromInterval } from ".";

/**
 * Function to move a student from the larget group to the smallest group.
 * It will additionally move one restriction violation to a random group.
 * This helps makes the groups even and remove violations.
 * @param solution the current solution to modify
 * @param num_groups number of groups to make
 * @param restrictions students who cannot be in the same group
 * @returns new solution
 **/
function moveFromLargeGroup(
  solution: Solution,
  restrictions: Restriction[],
  num_groups: number,
): Solution {
  let sol = deepCopy(solution); // Copying the solution so as to not mutate the original array

  //TODO: this is run twice per step. Could abstract it out for efficiency
  let studentToGroup = new Map<number, number>();
  for (let entry of sol) {
    studentToGroup.set(entry.student.id, entry.group);
  }

  // Look for a constraint violation
  for (let { student1, student2 } of restrictions) {
    if (studentToGroup.get(student1) === studentToGroup.get(student2)) {
      let index = randomIntFromInterval(
        0,
        sol.filter((item) => item.group !== studentToGroup.get(student1))
          .length - 1,
      );
      sol.find((s) => s.student.id === student1)!.group = sol[index].group;
    }
    // Move student1 to a different group
  }

  // Find the largest group
  let groupSizes: number[] = Array(num_groups).fill(0);
  for (let item of sol) {
    groupSizes[item.group]++;
  }
  let largestGroup = groupSizes.indexOf(Math.max(...groupSizes));

  // Find the smallest group
  let smallestGroup = groupSizes.indexOf(Math.min(...groupSizes));

  // Move a random student from the largest group to the smallest group
  let indices = sol.reduce(
    (acc: number[], solItem, index) =>
      solItem.group === largestGroup ? [...acc, index] : acc,
    [],
  );
  let index = randomIntFromInterval(0, indices.length - 1);
  sol[indices[index]].group = smallestGroup;

  return sol;
}

function randomMove(
  solution: Solution,
  restrictions: Restriction[],
  num_groups: number,
): Solution {
  const sol = deepCopy(solution);
  let index = randomIntFromInterval(0, sol.length - 1);
  let new_group = randomIntFromInterval(0, num_groups - 1);
  sol[index].group = new_group;
  return sol;
}

/**
 * Makes moves in the solution in order to progress to a better one.
 * @param solution the current solution to modify
 * @param num_groups number of groups to make
 * @param restrictions students who cannot be in the same group
 * @returns new solution
 **/
export function makeMove(
  solution: Solution,
  restrictions: Restriction[],
  num_groups: number,
): Solution {
  return moveFromLargeGroup(solution, restrictions, num_groups);
}

File: annealing/annealing/move.py
"""
This module contains functions for making moves in the search space.
"""
import random


def make_move(solution, restrictions, num_groups):
    """
    Make a random move.
    """
    sol = move_from_large_group(solution, restrictions, num_groups)
    return sol


def move_from_large_group(solution: list[int], restrictions, num_groups):
    """
    Move a random student from the largest group to the smallest group.
    """
    # Look for a constraint violation
    for student1, student2 in restrictions:
        if solution[student1] == solution[student2]:
            # Move student1 to a different group
            index = random.choice(
                [
                    group
                    for _, group in enumerate(solution)
                    if group != solution[student1]
                ]
            )
            solution[student1] = index

    # Find the largest group
    group_sizes = [solution.count(i) for i in range(num_groups)]
    largest_group = group_sizes.index(max(group_sizes))

    # Find the smallest group
    smallest_group = group_sizes.index(min(group_sizes))

    # Move a random student from the largest group to the smallest group
    index = random.choice(
        [i for i, group in enumerate(solution) if group == largest_group]
    )
    solution[index] = smallest_group

    return solution


def random_move(solution, restrictions, num_groups):
    """
    Move a random student to a random group.
    """
    index = random.randint(0, len(solution) - 1)
    new_group = random.randint(0, num_groups - 1)
    solution[index] = new_group
    return solution
